core/src/qos_service.rs

```diff
@@ -327,16 +328,50 @@ mod tests {
         let qos_service_1 = qos_service.clone();
         let qos_service_2 = qos_service.clone();
 
-        let th_1 = thread::spawn(move || {
-            qos_service_1.compute_transaction_costs(txs_1.iter(), false);
-        });
+        let th_1 = Builder::new()
+            .name("test-producer-1".to_string())
+            .spawn(move || {
+                debug!("thread 1 starts with {} txs", txs_1.len());
+                let tx_costs = qos_service_1.compute_transaction_costs(txs_1.iter(), false);
+                assert_eq!(txs_count, tx_costs.len());
+                debug!(
+                    "thread 1 done, generated {} count, see service count as {}",
+                    txs_count,
+                    qos_service_1
+                        .metrics
+                        .compute_cost_count
+                        .load(Ordering::Relaxed)
+                );
+            })
+            .unwrap();
+
+        let th_2 = Builder::new()
+            .name("test-producer-2".to_string())
+            .spawn(move || {
+                debug!("thread 2 starts with {} txs", txs_2.len());
+                let tx_costs = qos_service_2.compute_transaction_costs(txs_2.iter(), false);
+                assert_eq!(txs_count, tx_costs.len());
+                debug!(
+                    "thread 2 done, generated {} count, see service count as {}",
+                    txs_count,
+                    qos_service_2
+                        .metrics
+                        .compute_cost_count
+                        .load(Ordering::Relaxed)
+                );
+            })
+            .unwrap();
 
-        let th_2 = thread::spawn(move || {
-            qos_service_2.compute_transaction_costs(txs_2.iter(), false);
-        });
+        th_1.join().expect("qos service 1 panicked");
```

Comment:
Since we are joining on the threads -- so I do not think the reason is the thread was failed to launched. I think something else is causing the sporadic issue.  The compute_cost_count is changed in two places:

1.                 (
                    "compute_cost_count",
                    self.compute_cost_count.swap(0, Ordering::Relaxed) as i64,
                    i64
                ),

2.         self.metrics
            .compute_cost_count
            .fetch_add(txs_costs.len() as u64, Ordering::Relaxed);

I am guessing it could happen this way:

1. One thread is done with compute_transaction_costs and update the count to 2048.
2. The report thread kicks in and reset the value to 0
3. The other thread run compute_transaction_costs again and set it to 2048.
4. The test loads the 2048.

Changing to shorter txn list only changes the timing part I think.

Maybe we can turn off report in the tests?

    fn reporting_loop(running_flag: Arc<AtomicBool>, metrics: Arc<QosServiceMetrics>) {
        while running_flag.load(Ordering::Relaxed) {
            // hardcode to report every 1000ms
            metrics.report(1000u64);
            thread::sleep(Duration::from_millis(100));
        }
    }

Reply:
Sharp eyes! Thanks @lijunwangs , this is very likely cause, I totally overlooked reporting part. Putting in a fix soon